tools/c7n_gcp/c7n_gcp/actions/labels.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2019 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from c7n.utils import local_session, type_schema
+from c7n_gcp.actions import MethodAction
+from c7n.filters import FilterValidationError
+from c7n_azure.lookup import Lookup
+
+
+def register_labeling(action_registry):
+    action_registry.register('label', Label)
+    action_registry.register('unlabel', RemoveLabel)
+
+
+class BaseLabelAction(MethodAction):
+    schema = type_schema('setLabels')
+    method_spec = {'op': 'setLabels'}
+
+    def get_labels_to_add(self, resource):
+        return None
+
+    def get_labels_to_delete(self, resource):
+        return None
+
+    def _merge_labels(self, current_labels, new_labels, remove_labels):
+        result = current_labels
+        if new_labels:
+            result.update(new_labels)
+        if remove_labels:
+            result = {k: v for k, v in current_labels.items() if k not in remove_labels}
+        return result
+
+    def get_resource_params(self, model, resource):
+        params = model.get_self_params(resource)
+        current_labels = self._get_current_labels(resource)
+        new_labels = self.get_labels_to_add(resource)
+        remove_labels = self.get_labels_to_delete(resource)
+        all_labels = self._merge_labels(current_labels, new_labels, remove_labels)
+        body = {'body': {
+            'labels': all_labels,
+            'labelFingerprint': resource['labelFingerprint']
+        }}
+
+        params.update(body)
+        return params
+
+    def _get_body(self, resource):
+        pass
+
+    def _get_current_labels(self, resource):
+        return resource.get('labels', {})
+
+
+class Label(BaseLabelAction):
```

Comment:
given the underlying implementation is a set (add & delete), feels like we should model that instead of decomposing into add/remove actions.

Reply:
I left it with compatybility..

And from usage perspective.. its much easier to use `add` `remove` than `set`

Reply:
But I see that actually set would be much better..
Cause if you do add + remove at same resource it sometimes breaks..

Reply:
```
set-labels:
  add:  mapping/object
  remove: array of strings
```
should work okay. for some resources that present as a more document oriented, to avoid concurrency issues we'll sometimes refetch and merge, but modeling as a set avoids the need for multiple actions in a given policy which inherently conflict

Reply:
Yes but then its harder to add `mark-for-op` end things like `trim`, `auto-label` etc..
With current approach it would be much easier to add/extend..

Reply:
the other action usages can still use set-label via composition, the reuse pattern isn't significantly different or more difficult.

Reply:
Yes but we still need set label refresh state after/before im not sure?